### ISSUE TITLE
Maintaining Document IDs

### ISSUE THREAD

**Issue by shohil-kishore** [MaartenGr/BERTopic#790]
Hi Maarten,

Thanks for creating such a useful package! I have a sample of approx. five million tweets and I'm wondering if you know of a simple way to keep doc IDs after topic modelling has occurred? After topic modelling, I can extract all docs per topic, but without an ID, I'm unable to identify who posted the tweets, how many retweets the tweet got, etc. Any help is greatly appreciated! 

Thanks,
Shohil

**Comment by zachschillaci27** [comment]
Hi Shohil, I'll try to answer here.

I would recommend starting from a data table (eg. `pd.DataFrame`) where you have all the tweets stored along with their ID's and metadata. Then run the `BERTopic` fit + transform and store the output topic predictions (along with their probabilities if needed) as a new column in the table. As far as I know, the ordering of the model predictions should be aligned with the order of the data passed to it. This will give you a complete table containing the raw data and model predictions in one.

**Comment by shohil-kishore** [comment]
Hi Zach, thanks for the reply! That much makes sense to me but I’m not sure what variable contains an ordered list of the output topic predictions which can be merged as a new DF column - do you know by any chance? Thanks for your help.

**Comment by ceesroele** [comment]
Saying the same as zachschillaci27 above, but with some code:
AFAIK, the only "ordered list" is the one returned from `fit_transform`. There is no topic model variable that can be retrieved. Again AFAIK, the only way to get topic ids later (e.g. from a saved & loaded model)  is to call `transform` again.
```
df.columns = ['doc_id', 'text']  # Example: we have two columns in our dataset: `doc_id` to identify the record and `text`
docs = df['text'].tolist()
predictions, _ = topic_model.fit_transform(docs)  # I ignore probabilities here
df['topic_id] = predictions  # Here we add a new column to our data, indicating the topic_id
df.to_csv('myfile.csv')         # Save data with topic_id for later use
```

**Comment by MaartenGr** [comment]
@shohil-kishore There are multiple ways to connect the topics that were calculated with the input documents. As @ceesroele indicated, you can use the output from the `fit_transform` as they are ordered. However, there actually **is** a topic model variable that can be retrieved for that purpose, namely `topic_model.topics_`. In that variable, you will find the topic for each document in the same order as `docs`. In other words, you can then do the following:

```python
import pandas as pd
from bertopic import BERTopic

# We train our model
topics, probs = topic_model.fit_transform(docs) 

# Then we match our documents with our topics using `topics`:
df = pd.DataFrame({"Document": docs, "Topic": topics})

# Or we match our documents with our topics using `topic_model.topics_`
df = pd.DataFrame({"Document": docs, "Topic": topic_model.topics_})
```

In other words, there is no need to save the data with topic ID for later use.

**Comment by zachschillaci27** [comment]
@MaartenGr that's good to know! I wasn't aware of this variable. However, I would disagree with this point below:

> In other words, there is no need to save the data with topic ID for later use.

Because, as far as I know, there's no internal mapping  recording the documents passed to `fit` or `fit_transform`, no? In this case I would recommend always storing documents with their associated topics (either from the fit/transform return or `topic_model.topics_`) immediately after running the model inference. This will ensure that in case the docs are overwritten, shuffled, or for any other reason the original document ordering is not preserved, the topic predictions for the original documents may still be recovered.

**Comment by shohil-kishore** [comment]
@MaartenGr and @ceesroele - that's exactly what I was looking for, thank you! 👍 

Thanks also to @zachschillaci27 for the quick response!
